packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
+ * forwarded to a promise q in another vat. Once q is settled, given eventual
+ * connectivity and continued progress, r will eventually be settled. If
+ * q is rejected, r will eventually be rejected, either with the same reason
+ * or with a reason from the "system". If q is fulfulled, then r will
+ * eventually either be fulfilled with the result of passing the fulfullment
+ * (as if as an argument in a message), or r will be rejected with a reason
+ * from the "system". For example, if q's fulfillment is not passable, then
+ * r will be rejected with this non-passability as the reason. If q's
+ * fulfullment is a PrimaryRemotable T, then r must eventually fulfill to
+ * a Remote<T> with the same primary.
+ *
+ * If a promise is forwarded into a forwarding loop back to itself, whether
+ * local or distributed, all the promises in that loop must eventually be
+ * rejected.
+ *
+ * An eventual message sent to a promise that eventually fulfills must
+ * eventually be delivered to its fulfillment.
  */
 
 /**
- * @typedef {Passable} Comparable
+ * @template {Passable} T
+ * @typedef {T[]} CopyArray
+ * Corresponding PassStyle is "copyArray".
+ *
+ * The semantics of CopyArrays are modeled on Tuples from the Records and
+ * Tuples proposal. A CopyArray consists only of the sequence of its elements.
+ *
+ * CopyArray and CopyRecord are parameterized over various "deep" constraints
+ * explained below. For example, iff all the elements of a CopyArray are
+ * PassableKeys, then the CopyArray is a PassableKey that can be compared for
+ * equivalence with `sameKey`. Two such CopyArrays are equivalent if their
+ * elements are pairwise equivalent.
  *
- * A Comparable is a Passable in which none of the leaves of the pass-by-copy
- * superstructure are promises. Two Comparables may be compared by
- * for equivalence according to `sameStructure`, which is the strongest
- * equivalence class supported by marshal's distributed object semantics.
+ * A CopyArray as a JS object is a hardened plain array with no holes, no extra
+ * properties, the normal non-enumerable `length` property, and enumerable
+ * data properties from 0 to length-1.
  */
 
 /**
- * @typedef {Comparable} OnlyData
+ * @template {Passable} T
+ * @typedef {{name: string]: T}} CopyRecord
+ * Corresponding PassStyle is "copyRecord".
  *
- * A Comparable is OnlyData when its pass-by-copy superstructure has no leaves,
- * i.e., when all the leaves of the data structure tree are primitive data
- * types or empty composites.
+ * The semantics of CopyRecords is modeled on Records from the Records and
+ * Tuples proposal.
+ * Property names can only be strings. A CopyRecord consists only of
+ * a set of (property-name, value) pairs, where each property name
+ * can appear at most once. It is therefore a single-valued mapping
+ * from property-name to value. CopyRecords enumerate their properties
+ * according to the sort order of their property names.
```

Comment:
It's amusing that in-between the period where I last looked at Agoric's CapTP, both Goblins and Agoric added a dynamic record-like system.  The main differences IIUC:

- Agoric's/JS's is unlabeled and uses keyword-based arguments
- Goblins' is labeled and uses positional ordering

I wonder if both are worth supporting; these are both useful things to have.  It's interesting that these mirror the two choices for passing arguments to functions, which is not surprising when considering that they might generally mirror whatever arguments they would be passed into to some local constructor (or abstractly represent the same without doing so).

I am happy to support Agoric's record structure in Goblins.  I will probably call it "KRecords", to indicate "keyword-based records".  Maybe it would be appropriate to call Goblins' "PLRecords", meaning "positional, labeled records".

I suspect PLRecords don't have a usage-based priority in Agoric's code at the moment... I also think convincing you to adopt them is not the highest priority either, though I will commit to adding KRecords at the moment.  But I wanted to put it on the table as something to discuss.

Reply:
Also, I'd like to clarify... CopyRecord is meant to be distinct from presumably what will be called CopyMap.  Is this true?  We had some discussion of the "conceptual difference" between the two so I have operated off that assumption.

Reply:
It sounds like your KRecords are our CopyRecords. Good!

Yes, CopyMap is very distinct from CopyRecord. CopyRecord property names are only strings, which therefore can be sorted. If a CopyRecord contains only PassableKeys as values, then the CopyRecord itself is a PassableKey.

CopyMaps use PassableKeys as keys. But because PassableKeys cannot be sorted, CopyMaps themselves are not PassableKeys.

Echoing @kentonv's point about difference in purpose, a record normally has a statically known set of property names with distinctly typed values. A map normally has a dynamically computed set of keys and values, where there's generally typing uniformity among the keys and separately among the values.

Taking that to JS, records are written as object literals whereas maps are defined by APIs. We have no desire to support the JS objects-as-map antipattern.